File: python/sheldon.py

from possible_friend import PossibleFriend
from random import choices


class Sheldon:

    def __init__(self):
        pass

    def ask_for_meal(self, friend: PossibleFriend) -> bool:
        print('- Would you like to share a meal?')
        answer = friend.answer_yes_or_no()
        if answer == 'Yes':
            print('- Excellent. Let\'s dine together and begin our friendship!')
            return True
        else:
            return False

    def ask_for_drink(self, friend: PossibleFriend) -> bool:
        print('- Alright, do you enjoy a hot beverage?')
        answer = friend.answer_yes_or_no()
        if answer == 'Yes':
            print("- Excellent. Popular choices include tea, coffee, cocoa")
            drink = friend.posible_desirable_drink()
            if drink in ['tea', 'coffee', 'cocoa']:
                print(
                    f"- Let's have {drink} together and begin our friendship!")
                return True
            else:
                print(f'- I\'m sorry, I don\'t do {drink}')
                return False
        else:
            return False

    def ask_for_interests(self, friend: PossibleFriend) -> bool:
        self.interests = None
        howard_counter = 10
        collected_interests = []
        for i in range(howard_counter):
            print(f'- Tell me an interest of yours?')
            f_interest = friend.interests()
            collected_interests.append(f_interest)
            print(f"Really, {f_interest}?! I don't do {f_interest}.")
            if i > 6:
                least_objectionable_choice = choices(collected_interests)[0]
                print(
                    f'- Why don\'t we do that together? Let\'s partake in  {least_objectionable_choice} and begin our friendship!')
                return True

        return False

File: ts/sheldon.ts
/*
 * the BiG BANG THEORY
 * Season 2, Episode 13
 * The Friendship Algorithm
 */
export const createSheldon = () => {
    const shareMeal = function () {
        const letsEat = (prompt("Would you like to share a meal? Yes or no?") || "").toLowerCase();

        if (letsEat == 'yes') {
            const excellent = alert("Excellent. Let's dine together and begin our friendship!");
            return excellent;
        } else {
            shareDrink();
        }
    };

    const shareDrink = function () {
        const letsDrink = (prompt("Alright, do you enjoy a hot beverage? Yes or no?") || "").toLowerCase();

        if (letsDrink == 'yes') {
            const popChoices = (prompt("Excellent. Popular choices include tea, coffee, cocoa?") || "").toLowerCase();
            switch (popChoices) {
            case "tea":
                alert("Let's have " + popChoices + " together and begin our friendship!");
                break;
            case "coffee":
                alert("Let's have " + popChoices + " together and begin our friendship!");
                break;
            case "cocoa":
                alert("Let's have " + popChoices + " together and begin our friendship!");
                break;
            default:
                alert("I'm sorry , I don't do " + popChoices + ". What about a recreational activity? I bet we share some common interests!");
                shareInterest();
                break;
            }
        } else {
            alert("What about a recreational activity? I bet we share some common interests!");
            shareInterest();
        }
    };

    const shareInterest = function () {
        //Since Sheldon will say no to everything, this object is absolutely false.
        const sheldonsInterests = false;

        //Create the replied interests and the array it will be assigned to.
        let yourInterests;

        //Sheldon won't agree with you, period.
        if (yourInterests === sheldonsInterests) {
            //This doesn't need to be here but wishful thinking doesn't hurt.
            alert("Why don't we do that together? Let's partake in " + yourInterests + " and begin our friendship!");
        } else {

            //Create the array that the interests will be assigned to.
            const collectResponses: string[] = [];

            //Howard's loop counter.
            let n = 0;
            while (n < 10) {
                n++;
                yourInterests = (prompt("Tell me an interest of yours?") || "").toLowerCase();
                //Update the resonse and store it into the collection
                collectResponses.push(yourInterests);
                alert("Really, " + yourInterests + "?! I don't do " + yourInterests + ".");

                if (n > 6) {
                    //Randomly choose the least objectionable interest.
                    //I don't know what goes on in Shelly's mind so let's randomly choose.
                    const shellysRandomChoice = collectResponses[Math.floor(Math.random() * collectResponses.length)];
                    alert("Why don't we do that together? Let's partake in " + shellysRandomChoice + " and begin our friendship!");
                    break;
                }
            }
        }
    };

    return {
        ask: function () {
            shareMeal();
        }
    };
};
